File: 50dias/cpp/33.cpp
#include <iostream>
#include <fstream>
#include <algorithm>
#include <vector>
#include <list>
#include <string>

using namespace std;

const vector<char> vogais = { 'a', 'e', 'i', 'o', 'u',
							 'A', 'E', 'I', 'O', 'U' };

void LerArquivo(const string& nomeDoArquivo, list<string>& texto) {
	ifstream inputFile;
	inputFile.open(nomeDoArquivo);

	string linha;
	while (getline(inputFile, linha)) {
		texto.push_back(linha);
	}
	inputFile.close();
}

void EscreveArquivo(const string& nomeDoArquivo, const list<string>& texto) {
	ofstream outputFile;
	outputFile.open(nomeDoArquivo);

	for (auto linha : texto)
		outputFile << linha << endl;

	outputFile.close();
}

bool EhVogal(const char& letra)
{
	auto it = find(vogais.begin(), vogais.end(), letra);
	bool encontrou = it != vogais.end();
	return encontrou;
}

void CensuraTexto(
	const string& texto,
	string& textoCensurado,
	string& vogais
)
{
	for (int i = 0; i < texto.size(); i++) {
		if (EhVogal(texto[i])) {
			textoCensurado += '*';
			vogais += texto[i];
		}
		else {
			textoCensurado += texto[i];
		}
	}
}

string RemoveCensuraTexto(
	const string& textoCensurado,
	const string& vogais,
	int& contadorVogais)
{
	string textoOriginal;
	for (int i = 0; i < textoCensurado.size(); i++) {
		if (textoCensurado[i] == '*') {
			textoOriginal += vogais[contadorVogais];
			contadorVogais++;
		}
		else {
			textoOriginal += textoCensurado[i];
		}
	}
	return textoOriginal;
}

string CodificaTexto(const string& textoOriginal)
{
	string textoCodificado = "";
	for (auto letra : textoOriginal) {
		textoCodificado += letra + 3;
	}

	return textoCodificado;
}

string DecodificaTexto(const string& textoCodificado)
{
	string textoOriginal = "";
	for (auto letra : textoCodificado)
	{
		textoOriginal += letra - 3;
	}
	return textoOriginal;
}

void CodificaTextoCompleto(const list<string>& textoCompleto, list<string>& textoCodificado) {
	for (auto linha : textoCompleto) {
		textoCodificado.push_back(CodificaTexto(linha));
	}
}

void DecodificaTextoCompleto(const list<string>& textoCompleto, list<string>& textoDecodificado) {
	for (auto linha : textoCompleto) {
		textoDecodificado.push_back(DecodificaTexto(linha));
	}
}

void CensuraTextoCompleto(
	const list<string>& textoCompleto,
	list<string>& textoCensurado,
	string& vogais
)
{
	for (auto linha : textoCompleto) {
		string linhaCensurada;
		CensuraTexto(linha, linhaCensurada, vogais);
		textoCensurado.push_back(linhaCensurada);
	}
}

void RemoveCensuraTextoCompleto(
	const list<string>& textoCensurado,
	list<string>& textoCompleto,
	string& vogais
)
{
	int contadorVogais = 0;
	for (auto linhaCensurada : textoCensurado) {
		string linha = RemoveCensuraTexto(linhaCensurada, vogais, contadorVogais);
		textoCompleto.push_back(linha);
	}
}

void CodificaArquivoTexto() {
	list<string> texto, textoCensurado, textoCodificado, chave;
	string vogais;

	LerArquivo("entrada.txt", texto);
	CodificaTextoCompleto(texto, textoCodificado);
	CensuraTextoCompleto(textoCodificado, textoCensurado, vogais);
	chave.push_back(vogais);

	EscreveArquivo("cod.txt", textoCodificado);
	EscreveArquivo("chave.txt", chave);
}

void DecodificaArquivoTexto() {
	list<string> texto, textoCensurado, textoCodificado, chave;
	string vogais;
	LerArquivo("cod.txt", textoCensurado);
	LerArquivo("chave.txt", chave);

	vogais = chave.front();

	RemoveCensuraTextoCompleto(textoCensurado, textoCodificado, vogais);
	DecodificaTextoCompleto(textoCodificado, texto);

	EscreveArquivo("saida.txt", texto);
}

int main()
{
	CodificaArquivoTexto();

	DecodificaArquivoTexto();
	return 0;
}

File: 50dias/python/33.py

def LeTextoDoArquivo(nomeDoArquivo):
    input_file = open(nomeDoArquivo, 'r')
    texto = input_file.read()
    input_file.close()
    return texto

def EscreveTextoNoArquivo(texto, nomeDoArquivo):
    output_file = open(nomeDoArquivo, 'w')
    output_file.write(texto)
    output_file.close()

def EhVogal(letra):
    return letra in ['a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U']

def CensuraTexto(texto):
    textoCensurado = ""
    vogais = ""
    for i in range(len(texto)):
        if EhVogal(texto[i]):
            textoCensurado += '*'
            vogais += texto[i]
        else:
            textoCensurado += texto[i]
    return textoCensurado, vogais

def RemoveCensuraTexto(textoCensurado, vogais):
    contadorVogais = 0
    textoOriginal = ""
    for i in range(len(textoCensurado)):
        if textoCensurado[i] == '*':
            textoOriginal += vogais[contadorVogais]
            contadorVogais += 1
        else:
            textoOriginal += textoCensurado[i]
    return textoOriginal

def CodificaTexto(textoOriginal):
    textoCodificado = ""
    for letra in textoOriginal:
        if letra == '\n':
            textoCodificado += letra
        else:
            textoCodificado += chr(ord(letra) + 3)
    return textoCodificado

def DecodificaTexto(textoCodificado):
    textoOriginal = ""
    for letra in textoCodificado:
        if letra == '\n':
            textoOriginal += letra
        else:
            textoOriginal += chr(ord(letra) - 3)
    return textoOriginal

def EfetuaCodificacaoDoArquivo(texto):
    textoCodificado = CodificaTexto(texto)
    textoCensurado, vogais = CensuraTexto(textoCodificado)

    EscreveTextoNoArquivo(textoCensurado, 'cod.txt')
    EscreveTextoNoArquivo(vogais, 'chave.txt')

def DecodificaTextos():
    textoCensurado = LeTextoDoArquivo('cod.txt')
    vogais = LeTextoDoArquivo('chave.txt')
    textoSemCensura = RemoveCensuraTexto(textoCensurado, vogais)
    textoOriginal = DecodificaTexto(textoSemCensura)
    return textoOriginal

def main():
    texto = LeTextoDoArquivo('entrada.txt')

    EfetuaCodificacaoDoArquivo(texto)
    textoDecodificado = DecodificaTextos()

    EscreveTextoNoArquivo(textoDecodificado, 'saida.txt')


main()



















